File: Binary_Tree_Postorder_Traversal/Binary_Tree_Postorder_Traversal.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def postorder_traversal(self, root):
        ans = []
        if not root:
            return ans

        stack = []
        last_visited = None

        while root or stack:
            while root:
                stack.append(root)
                root = root.left

            node = stack[-1]

            if node.right and last_visited != node.right:
                root = node.right
            else:
                ans.append(node.val)
                last_visited = stack.pop()

        return ans

def build_tree(values):
    if not values:
        return None

    nodes = [TreeNode(val) if val is not None else None for val in values]
    kids = nodes[::-1]
    root = kids.pop()

    for node in nodes:
        if node:
            if kids:
                node.left = kids.pop()
            if kids:
                node.right = kids.pop()

    return root

values = [1, None, 2, 3]
root = build_tree(values)
print(Solution().postorder_traversal(root))

values2 = [1,2,3,4,5,None,8,None,None,6,7,9]
root2 = build_tree(values2)
print(Solution().postorder_traversal(root2))

values3 = []
root3 = build_tree(values3)
print(Solution().postorder_traversal(root3))

values4 = [1]
root4 = build_tree(values4)
print(Solution().postorder_traversal(root4))

File: Binary_Tree_Postorder_Traversal/Binary_Tree_Postorder_Traversal.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();

        if (root == null) {
            return ans;
        }
        TreeNode lastVisited = null;

        while (root != null || !stack.isEmpty()) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }

            TreeNode node = stack.peek();

            if (node.right != null && lastVisited != node.right) {
                root = node.right;
            } else {
                ans.add(node.val);
                lastVisited = stack.pop();
            }
        }

        return ans;
    }
}

public class Main {
    public static TreeNode buildTree(Integer[] values) {
        if (values == null || values.length == 0) {
            return null;
        }
        TreeNode[] nodes = new TreeNode[values.length];
        for (int i = 0; i < values.length; i++) {
            if(values[i] != null) {
                nodes[i] = new TreeNode(values[i]);
            }
        }

        int index = 1;
        for (int i = 0; i < values.length; i++) {
            if (nodes[i] != null) {
                if (index < values.length && nodes[index] != null) {
                    nodes[i].left = nodes[index];
                }
                index++;
                if (index < values.length && nodes[index] != null) {
                    nodes[i].right = nodes[index];
                }
                index++;
            }
        }
        return nodes[0];
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        Integer[] values1 = {1, null, 2, 3};
        TreeNode root1 = buildTree(values1);
        System.out.println(solution.postorderTraversal(root1));

        Integer[] values2 = {1, 2, 3, 4, 5, null, 8, null, null, 6, 7, 9};
        TreeNode root2 = buildTree(values2);
        System.out.println(solution.postorderTraversal(root2));

        Integer[] values3 = {};
        TreeNode root3 = buildTree(values3);
        System.out.println(solution.postorderTraversal(root3));

        Integer[] values4 = {1};
        TreeNode root4 = buildTree(values4);
        System.out.println(solution.postorderTraversal(root4));
    }
}